go/vt/vtgate/planbuilder/operators/query_planning.go

```diff
@@ -215,23 +220,100 @@ func pushOrExpandHorizon(ctx *plancontext.PlanningContext, in *Horizon) (Operato
 	return expandHorizon(ctx, in)
 }
 
-func tryPushLimit(in *Limit) (Operator, *ApplyResult) {
+func tryPushLimit(ctx *plancontext.PlanningContext, in *Limit) (Operator, *ApplyResult) {
 	switch src := in.Source.(type) {
 	case *Route:
-		return tryPushingDownLimitInRoute(in, src)
+		return tryPushingDownLimitInRoute(ctx, in, src)
 	case *Aggregator:
 		return in, NoRewrite
+	case *ApplyJoin:
+		if in.Pushed {
+			// This is the Top limit, and it's already pushed down
+			return in, NoRewrite
+		}
+		side := "RHS"
+		src.RHS = createPushedLimit(ctx, src.RHS, in)
+		if IsOuter(src) {
+			// for outer joins, we are guaranteed that all rows from the LHS will be returned,
+			// so we can push down the LIMIT to the LHS
+			src.LHS = createPushedLimit(ctx, src.LHS, in)
+			side = "RHS and LHS"
+		}
+
+		if in.Top {
+			in.Pushed = true
+			return in, Rewrote(fmt.Sprintf("add limit to %s of apply join", side))
+		}
```

Comment:
I think we need to keep the Limit on top of a Join even if its not the top limit. If the Join is going to be an input for another operator, and there is a limit on that join, then just pushing it to either side is not enough. After doing the join too, it is a good idea to limit the rows again. For example look at the following plan - 
```json
{
            "Instructions": {
                "Count": "5",
                "Inputs": [
                    {
                        "Inputs": [
                            {
                                "FieldQuery": "select 1 from user_extra where 1 != 1",
                                "Keyspace": {
                                    "Name": "user",
                                    "Sharded": true
                                },
                                "OperatorType": "Route",
                                "Query": "select 1 from user_extra",
                                "Table": "user_extra",
                                "Variant": "Scatter"
                            },
                            {
                                "Inputs": [
                                    {
                                        "FieldQuery": "select 1 from `user` where 1 != 1",
                                        "Keyspace": {
                                            "Name": "user",
                                            "Sharded": true
                                        },
                                        "OperatorType": "Route",
                                        "Query": "select 1 from `user`",
                                        "Table": "`user`",
                                        "Variant": "Scatter"
                                    },
                                    {
                                        "FieldQuery": "select unsharded_a.x from unsharded_a where 1 != 1",
                                        "Keyspace": {
                                            "Name": "main",
                                            "Sharded": false
                                        },
                                        "OperatorType": "Route",
                                        "Query": "select unsharded_a.x from unsharded_a limit 5",
                                        "Table": "unsharded_a",
                                        "Variant": "Unsharded"
                                    }
                                ],
                                "JoinColumnIndexes": "R:0",
                                "OperatorType": "Join",
                                "TableName": "`user`_unsharded_a",
                                "Variant": "Join"
                            }
                        ],
                        "JoinColumnIndexes": "R:0",
                        "OperatorType": "Join",
                        "TableName": "user_extra_`user`_unsharded_a",
                        "Variant": "Join"
                    }
                ],
                "OperatorType": "Limit"
            },
            "Original": "select unsharded_a.x from user join user_extra join unsharded_a limit 5",
            "QueryType": "SELECT",
            "TablesUsed": [
                "main.unsharded_a",
                "user.user",
                "user.user_extra"
            ]
        }
```
I feel having a limit on top of the inner joins is still beneficial.